gh-117482: Cleaning up tp_dict for static builtin types in subinterpreters

mliezun opened the pull request:
Solves #117482 

## Problem description

When initializing base types in a subinterpreters extra items are added to tp_dict. For example, in the linked issue we use the following script to identify the problem:

```python
import _xxsubinterpreters as interpreters

script = """print(int.__str__)"""


exec(script)
# Output: <slot wrapper '__str__' of 'object' objects>

interp_id = interpreters.create()
interpreters.run_string(interp_id, script)
# Output: <slot wrapper '__str__' of 'int' objects>
```

While running on a subinterpreter `int.__str__` points to a slot in `int` itself instead of pointing to the one defined in `object` as it should be.

## Tracking down the root cause

As can be seen in https://github.com/python/cpython/issues/117482#issuecomment-2033124539, the problem was introduced in commit https://github.com/python/cpython/commit/de64e7561680fdc5358001e9488091e75d4174a3.

By further debugging the specific parts modified in that commit I found that the problem lies in the [type_ready](https://github.com/python/cpython/blob/df73179/Objects/typeobject.c#L7821) function.

In previous versions of Python `tp_dict` was shared between subinterpreters. Now each interpreter has their own copy for static builtin types.

Let's take `int.__str__` as an example to see what's happening:

1. The `type_ready` function gets executed as part of the main interpreter initialization process.
2. The function starts "readying" the `int` type: `tp_dict` gets filled by [type_ready_fill_dict](https://github.com/python/cpython/blob/df73179/Objects/typeobject.c#L7860), one of the thing this function does is lookup which slots are not empty and add them to the dict, for `int` the `tp_str` slot [is empty](https://github.com/python/cpython/blob/df73179/Objects/longobject.c#L6555).
3. After that, [type_ready_inherit](https://github.com/python/cpython/blob/df73179/Objects/typeobject.c#L7864C13-L7864C31) get called, and copies the [tp_str slot from object to int](https://github.com/python/cpython/blob/df73179/Objects/typeobject.c#L7274).
4. The init process for the main interpreter finishes and we're ready to start with the second interpreter.
5. In this case, the `tp_str` slot is not empty for int, it was filled by step 3. So it gets added to the dict by `type_ready_fill_dict`.
6. The program continues and we see different behavior depending on which interpreter we run.

The problem here is that the code expects `type_ready_fill_dict` to be called before `type_ready_inherit`. Which is true for the main interpreter but not for the subinterpreter. Also, this is caused because all the slots in `int` are shared except for `tp_dict` which is stored in each interpreter.

## Tracking down all affected types

I ran a little script to track down all affected types and the differences in their base dicts:

```python
import sys
import subprocess

type_names = [b for b in dir(__builtins__) if getattr(eval(b), '__dict__', None)]
for tp in type_names:
    script = f"print(set({tp}.__dict__.keys()))"

    process = subprocess.Popen([sys.executable, '-c', script], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, _ = process.communicate()
    original_keys = eval(out)

    script_interpreters = f"""
import _xxsubinterpreters as interpreters

interp_id = interpreters.create()
interpreters.run_string(interp_id, '{script}')
interpreters.destroy(interp_id)
    """
    process = subprocess.Popen([sys.executable, '-c', script_interpreters], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, _ = process.communicate()
    new_keys = eval(out)


    print(f"{tp}:", (original_keys-new_keys), (new_keys-original_keys))
```

This was used to compare outputs with previous python versions (<3.12) and to check if the proposed solution worked correctly.

## Proposed Solution

The current proposed solution is to add a function called `cleanup_tp_dict` that checks if we're running in a subinterpreter, then identifies what extra keys were added to the static type `tp_dict` and removes them.

This function is called at the end of `type_ready_fill_dict`.

## Alternative solutions

- Introduce new sentinel values similar `PyObject_HashNotImplemented` to indicate that certain slots should not be filled, like `tp_str` for `int`.
- Only call `type_ready_fill_dict` on the main interpreter and just make a copy of the original one in subinterpreters.

## Conclusion

The proposed solution feels quite hacky, but gets the work done. This is my first contribution to CPython and I don't have much experience with it so there surely might be a better way to do it, there's probably some internal knowledge that I'm missing.

I'm happy to work on a different approach if pointed in the right direction.

Thank you for your time and all your effort building Python!


<!-- gh-issue-number: gh-117482 -->
* Issue: gh-117482
<!-- /gh-issue-number -->

mliezun committed:
Cleaning up tp_dict for static builtin types in subinterpreter
When initializing base types in a subinterpreters extra items are added to tp_dict which triggers incorrect behavior while running code under a subinterpreter.
Here we're identifying which keys are not part of the original tp_dict in the main interpreter and removing them.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

ericsnowcurrently commented on Misc/NEWS.d/next/Core and Builtins/2024-06-03-23-13-29.gh-issue-117482.p5mtQI.rst:
```diff
@@ -0,0 +1 @@
+Fix builtins behavior on subinterpreters.
```
```suggestion
Fix "tp" slot wrappers for builtin types on subinterpreters.
In particular, this fixes a regression in methods like ``int.__str__()``.
```

ericsnowcurrently commented on Objects/typeobject.c:
```diff
@@ -279,6 +279,47 @@ lookup_tp_dict(PyTypeObject *self)
     return self->tp_dict;
 }
 
+static inline int
```
```suggestion
static int
```

ericsnowcurrently commented on Objects/typeobject.c:
```diff
@@ -7858,6 +7899,9 @@ type_ready_fill_dict(PyTypeObject *type)
     }
+    if (cleanup_tp_dict(type) < 0) {
+        return -1;
+    }
```
Let's move the top-level logic (and explanation) here, to simplify the new helper function.

```suggestion
    /* For now, we work around issues with slot wrappers
       for builtin types in subinterpreters.
       See https://github.com/python/cpython/issues/117482. */
    if (type->tp_flags & _Py_TPFLAGS_STATIC_BUILTIN) {
        PyInterpreterState *interp = _PyInterpreterState_GET();
        if (!_Py_IsMainInterpreter(interp)) {
            if (fix_builtin_slot_wrappers(type, interp) < 0) {
                return -1;
            }
        }
    }
```

ericsnowcurrently commented on Objects/typeobject.c:
```diff
@@ -279,6 +279,47 @@ lookup_tp_dict(PyTypeObject *self)
+        }
+    }
+    return 0;
+}
```
As noted in my other comment, let's move the top-level logic out.

```suggestion
fix_builtin_slot_wrappers(PyTypeObject *self, PyInterpreterState *interp)
{
    assert(self->tp_flags & _Py_TPFLAGS_STATIC_BUILTIN);
    assert(_Py_IsMainInterpreter(interp));

    static_builtin_state *state = _PyStaticType_GetState(interp, self);
    assert(state != NULL);
    PyInterpreterState *main_interp = _PyInterpreterState_Main();
    static_builtin_state *main_state = _PyStaticType_GetState(main_interp, self);
    assert(main_state != NULL);

    int res = -1;
    PyObject* keys_to_remove = PyList_New(0);
    if (keys_to_remove == NULL) {
        goto finally;
    }
    Py_ssize_t i = 0;
    PyObject *key, *value;
    while (PyDict_Next(state->tp_dict, &i, &key, &value)) {
        if (!PyDict_Contains(main_state->tp_dict, key)) {
            if (PyList_Append(keys_to_remove, key) < 0) {
                goto finally;
            }
        }
    }

    Py_ssize_t list_size = PyList_Size(keys_to_remove);
    for (Py_ssize_t i = 0; i < list_size; i++) {
        PyObject* key = PyList_GetItem(keys_to_remove, i);
        if (PyDict_DelItem(state->tp_dict, key) < 0) {
            goto finally;
        }
    }

    res = 0;
    
finally:
    Py_XDECREF(keys_to_remove);
    return res;
}
```

ericsnowcurrently commented on Objects/typeobject.c:
```diff
@@ -7858,6 +7899,9 @@ type_ready_fill_dict(PyTypeObject *type)
     if (type_dict_set_doc(type) < 0) {
         return -1;
     }
+    if (cleanup_tp_dict(type) < 0) {
```
Instead of adding a new helper function, I'd rather we fix the logic in the other functions called here in `type_ready_fill_dict()`.  It may make sense to merge this change first though.  (We have to factor in the risks of backporting any changes.)

ericsnowcurrently replied:
We should also add a test.  Something based on your reproducer, like the following, should work (in Lib/test/test_types.py):

```python
class SubinterpreterTests(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        global interpreters
        try:
            from test.support import interpreters
        except ModuleNotFoundError:
            raise unittest.Skip('subinterpreters required')
        import test.support.interpreters.channels

    @cpython_only
    def test_slot_wrappers(self):
        rch, sch = interpreters.channels.create()

        # For now it's sufficient to check int.__str__.
        # See https://github.com/python/cpython/issues/117482
        # and https://github.com/python/cpython/pull/117660.
        script = '''if True:
            text = repr(int.__str__)
            sch.send_nowait(text)
            '''

        exec(script)
        expected = rch.recv()

        interp = interpreters.create()
        interp.exec('from test.support import interpreters')
        interp.prepare_main(sch=sch)
        interp.exec(script)
        results = rch.recv()

        self.assertEqual(results, expected)
```

mliezun replied:
Thanks for the suggestion. I added the tests where you proposed and confirmed that it's working as expected.

ericsnowcurrently reviewed (changes requested):
I've left a few recommendations.

mliezun committed:
Update Misc/NEWS.d/next/Core and Builtins/2024-06-03-23-13-29.gh-issue-117482.p5mtQI.rst

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

mliezun committed:
Update Objects/typeobject.c

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

mliezun committed:
Update Objects/typeobject.c

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

mliezun committed:
Update Objects/typeobject.c

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

ericsnowcurrently committed:
Merge branch 'main' into fix-issue-117482

mliezun committed:
Fix linting by running: "pre-commit run --all-files"

mliezun committed:
Fix interpreter state data type

mliezun committed:
Corrected assert in fix_builtin_slot_wrappers
We only need to cleanup the dict slots when we're in a subinterpreter.

mliezun committed:
Test types in subinterpreters
Slot wrappers should behave the same in the main interpreter as well as in all subinterpreters.

mliezun committed:
Fix test for slot wrappers inside subinterpreters

mliezun committed:
Moved subinterpreter builtin types cleanup code to add_operators()

mliezun committed:
Skip slot wrapper test when subinterpreters are not available

mliezun committed:
Use SkipTest instead of skip
